clever/src/led.cpp

```diff
@@ -0,0 +1,300 @@
+			set_leds.request.leds[i].b = b;
+		}
+		call_set_leds();
+
+	} else if (current_effect.effect == "rainbow") {
+		for (int i = 0; i < led_count; i++) {
+			int pos = (int)round(counter + (255.0 * i / led_count)) % 255;
+			rainbow(pos % 255, r, g, b);
+			set_leds.request.leds[i].index = i;
+			set_leds.request.leds[i].r = r;
+			set_leds.request.leds[i].g = g;
+			set_leds.request.leds[i].b = b;
+		}
+		call_set_leds();
+	}
+}
+
+bool setEffect(clever::SetLEDEffect::Request& req, clever::SetLEDEffect::Response& res)
+{
+	res.success = true;
+	counter = 0;
+
+	if (req.effect != "flash" && current_effect.effect == req.effect &&
+	    current_effect.r == req.r && current_effect.g == req.g && current_effect.b == req.b) {
+		res.message = "Effect already set, skip";
+		return true;
+	}
+
+	if (req.effect == "" || req.effect == "fill") {
+		fill(req.r, req.g, req.b);
+
+	} else if (req.effect == "blink") {
+		timer.setPeriod(ros::Duration(1 / blink_rate), true);
+		timer.start();
+
+	} else if (req.effect == "blink_fast") {
+		timer.setPeriod(ros::Duration(1 / blink_fast_rate), true);
+		timer.start();
+
+	} else if (req.effect == "fade") {
+		timer.setPeriod(ros::Duration(0.05), true);
+		timer.start();
+
+	} else if (req.effect == "flash") {
+		ros::Duration delay(flash_delay);
+		fill(0, 0, 0);
+		delay.sleep();
+		fill(req.r, req.g, req.b);
+		delay.sleep();
+		fill(0, 0, 0);
+		delay.sleep();
+		fill(req.r, req.g, req.b);
+		delay.sleep();
+		fill(0, 0, 0);
+		delay.sleep();
+		if (current_effect.effect == "fill" || current_effect.effect == "fade") {
+			// restore previous filling
+			for (int i = 0; i < led_count; i++) {
+				fill(current_effect.r, current_effect.g, current_effect.b);
+			}
+			call_set_leds();
+		}
+		return true; // this effect happens only once
+
+	} else if (req.effect == "rainbow_fill") {
+		timer.setPeriod(ros::Duration(rainbow_period / 255), true);
+		timer.start();
+
+	} else if (req.effect == "rainbow") {
+		timer.setPeriod(ros::Duration(rainbow_period / 255), true);
+		timer.start();
+
+	} else {
+		res.message = "Unknown effect: " + req.effect + ". Available effects are fill, fade, blink, blink_fast, flash, rainbow, rainbow_fill.";
+		ROS_ERROR("%s", res.message.c_str());
+		res.success = false;
+		return true;
+	}
+
+	// set current effect
+	current_effect = req;
+	start_state = state;
+	start_time = ros::Time::now();
+
+	return true;
+}
+
+void handleState(const led_msgs::LEDStateArray& msg)
+{
+	state = msg;
+}
+
+bool notify(const std::string& event)
+{
+	if (ros::param::has("~notify/" + event + "/effect") ||
+	    ros::param::has("~notify/" + event + "/r") ||
+	    ros::param::has("~notify/" + event + "/g") ||
+	    ros::param::has("~notify/" + event + "/b")) {
+		ROS_INFO_THROTTLE(5, "led: notify %s", event.c_str());
+		clever::SetLEDEffect effect;
+		ros::param::get("~notify/" + event + "/effect", effect.request.effect);
+		int r, g, b;
+		ros::param::get("~notify/" + event + "/r", r);
```

Comment:
This is likely to fail if one of the fields is missing. Either your logical expression is wrong or these lines are. Consider using [the `param` call](http://docs.ros.org/kinetic/api/roscpp/html/namespaceros_1_1param.html#a903d64f2071733c1fb083aa22882bd20) to supply default values.

Reply:
No this is not failing. `ros::param::get` just returns `false` keeping the initial value of the variable (`0` for integers, `""` for strings).

Reply:
Your r, g, b values are not getting initialized then. Please change this to `ros::param::param`.